dockertest/output/dockerinfo.py

```diff
@@ -0,0 +1,104 @@
+    """
+    Parser of 'docker info' output
+    """
+
+    def __init__(self, info_string=None, docker_path=None):
+        if info_string is None:
+            if docker_path is None:
+                docker_path = 'docker'
+            info_string = subprocess.check_output(docker_path + ' info',
+                                                  shell=True,
+                                                  close_fds=True)
+        self.info_string = info_string
+        self.table = self._build_table(info_string)
+
+    @staticmethod
+    def _build_table(info_string):
+        """
+        'docker info' returns a human-readable list of key: value pairs.
+        Some are indented by a space, indicating that these key/values
+        are subelements of a previous element. To wit:
+
+             Images: 3
+             Server Version: 1.12.6
+             Storage Driver: devicemapper
+              Pool Name: vg--docker-docker--pool
+              Pool Blocksize: 524.3 kB
+              ...
+             Logging Driver: journald
+
+        In this case 'Images', 'Server Version', and 'Logging Driver'
+        are simple tuples but 'Storage Driver' has both a value and
+        a set of further key/value pairs: Pool Name, Pool Blocksize, ...
+
+        We parse that and create a dict with the expected key/value
+        mapping *and* an extra: for all elements with subelements,
+        'element...' (element name plus three dots) is a dict containing
+        the subelements. E.g. x['Storage Driver...']['Pool Name'] = 'vg--etc'
+        """
+        table = {}
+        current_key = None
+        for line in info_string.splitlines():
+            # Almost every line will be Foo: Bar, but 'Insecure Registries:'
+            # is followed by a simple list of IPv4 netmasks
+            if ': ' in line or line.endswith(':'):
+                (key, value) = [e.strip() for e in line.split(':', 1)]
+            else:
+                key = line.strip()
+                value = ''
+            if line.startswith(' '):
+                if not current_key:
+                    raise IndexError("Internal error: indented output line"
+                                     " '%s' from docker info with no previous"
+                                     " unindented lines.", line)
+                if current_key not in table:
+                    table[current_key] = {}
+                table[current_key][key] = value
+            else:
+                table[key] = value
+                current_key = key + '...'
+        return table
+
+    def __getitem__(self, key):
+        """
+        Each 'docker info' field has a scalar value; some also have
+        a further set of key/value pairs. Return both.
+        """
+        if key in self.table:
+            return self.table[key]
+        return None
+
+    def sub_fields(self, key):
+        """
+        Fetcher for multi-valued docker info elements. Always returns
+        a dict, even if empty.
```

Comment:
When you took notice of the excessive number of comments/docs needed for some part of ADEPT, it raised the "too complex" red-flag for me.  Take a step back, look at how much documenting is needed here, and think how the code can be simpler.

> Key and Value are strings. Always

yet

> Some docker info ... set of further Subkey

There are no overlapping keys, so why not just represent it that way?  e.g.

```python
def build_table(self):
    table = {}
    for line in self.info_string.splitlines():
        if ': ' in line or line.endswith(':'):
            (key, value) = [e.strip() for e in line.split(':', 1)]
        else:
            key = line.strip()
            value = ''
        key = key.lower().replace(' ', '_')  # All same-same
        if key in table:
            raise KeyError("Duplicate key %s detected, shoot Chris" % key)
        table[key] = value
    return table
```

Then, if you want to make callers life easier for some cases, add a method (or function):

```python
dm_keys = ('pool_name', 'pool_blocksize', ...)

def devicemapper_table(self):
    return dict(zip([(k, self[k]) for k in self.dm_keys]))
```

If in the future I get shot...well...then I won't really much care what happens from the grave 😀